### PR TITLE
Implement a simple DetermineProviderLocation

### LINKED ISSUE #14339: OTA Requestor needs to support querying providers from multiple fabrics
#### Problem
https://github.com/project-chip/connectedhomeip/issues/13782 will add support to periodic query using the `DefaultOTAProviders` attribute. The `DefaultOTAProviders` attribute is a list of default providers for each fabric that the requestor has been commissioned on. Currently, only one provider will ever be queried at a given time. Based on that assumption, there should be solution on which provider/fabric to query whenever the periodic timer fires.

 #### Proposed Solution
One solution could be that if a CASE session cannot be established with a particular provider, then the provider at the next fabric in the list should be queried. Perhaps a cache of which providers should be skipped/ignored needs to be maintained to support this.

There could be more complicated solutions such as using the `status` field of the QueryImageResponse to determine if a different provider should be used for query.

### PR DESCRIPTION
**Pull request by harsha-rajendran**
#### Problem
Provide a generic implementation for DetermineProviderLocation()

#### Change overview
It has been decided that a simple "iterate-through-the-list" approach would be the best path forward.
DetermineProviderLocation() iterates through the list of DefaultOTAProviders, finds the last-used-provider and returns the entry after it (cycles back to the start of the list). If no suitable candidate has been found, an error is returned.

#### Testing
- Tested by verifying the OTARequestor runs through the entire list in order (by reducing the periodicity of the QueryImage timer)
- Ensured OTA download works on Linux (With #15963 changes)

### CODE REVIEW

**Review by isiu-apple — APPROVED** [review]

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.h [hunk 1]
```diff
@@ -87,7 +87,7 @@ class GenericOTARequestorDriver : public OTARequestorDriver
     uint32_t mPeriodicQueryTimeInterval = (24 * 60 * 60); // Timeout for querying providers on the default OTA provider list
 
     using ProviderLocationType = app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type;
-    ProviderLocationType mLastUsedProvider; // Provider location used for the last query or update
+    chip::Optional<ProviderLocationType> mLastUsedProvider; // Provider location used for the last query or update
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.h:90]: nit: This is already wrapped in the chip namespace
```suggestion
   Optional<ProviderLocationType> mLastUsedProvider; // Provider location used for the last query or update
```

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 2]
```diff
@@ -320,17 +321,48 @@ void GenericOTARequestorDriver::StopDefaultProviderTimer()
...
-
+    else if (mLastUsedProvider.HasValue())
+    {
+        providerLocation = mLastUsedProvider.Value();
+        return true;
+    }
+    else
+    {
+        ChipLogError(SoftwareUpdate,
+                     "GenericOTARequestorDriver::DetermineProviderLocation: No suitable OTA Provider candidate found");
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:363]: nit: I think we try to avoid printing the function/file names out, e.g. don't use __FUNC__, etc

Comment (Damian-Nordic) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:363]: Right, it unnecessarily increases flash usage, so it's better to keep logs short and concise.

File: src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp [hunk 3]
```diff
@@ -68,9 +68,16 @@ void ExtendedOTARequestorDriver::PollUserConsentState()
...
-    subject.providerEndpointId = mLastUsedProvider.endpoint;
+    if (mLastUsedProvider.HasValue())
+    {
+        // mLastUsedProvider has the provider fabric index and endpoint id
+        subject.fabricIndex        = mLastUsedProvider.Value().fabricIndex;
+        subject.providerEndpointId = mLastUsedProvider.Value().endpoint;
+    }
+    else
+    {
+        ChipLogError(SoftwareUpdate, "ExtendedOTARequestorDriver::GetUserConsentSubject: mLastProvider is empty");
```
Comment (Damian-Nordic) [inline src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp:79]: Is the user of this function happy with fabricIndex and providerEndpointId not being initialized? Shouldn't we simply error in such a case?

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp [hunk 4]
```diff
@@ -320,17 +321,48 @@ void GenericOTARequestorDriver::StopDefaultProviderTimer()
...
+    // - error (no suitable candidate)
+    iterator = mRequestor->GetDefaultOTAProviderListIterator();
+    if (iterator.Next())
     {
-        // For now, just return the first one
         providerLocation = iterator.GetValue();
         return true;
     }
-
+    else if (mLastUsedProvider.HasValue())
```
Comment (Damian-Nordic) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:355]: Why is this needed? It can only be true if the last used provider is no longer among the default OTA providers. I don't think we should keep using it, right?

File: src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp (hunk 3 above)
Comment (hrajendran-samsung) [reply src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp:79]: Agreed- unless there's some convoluted way where the mLastUsedProvider was not set before sending the QueryImage out.
I see that OTARequestor::TriggerImmediateQuery() does not set the mLastUsedProvider, maybe there are others, that should be fixed?

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 2 above)
Comment (hrajendran-samsung) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:363]: Ok, I can change that. Regarding the point about increased RAM usage, we can have compile-time options to change the log-level and thus remove these constant string symbols, no?

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 4 above)
Comment (hrajendran-samsung) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:355]: So, this should only happen in the case where the Default OTA Providers List is empty for some reason. This should not happen unless something bad has happened (like an entity wiping the OTA providers list or a bug). In this case, I thought returning the only "valid" OTA provider was better than nothing at all. For all we know this provider might have an update.

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 2 above)
Comment (Damian-Nordic) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:363]: yes, but this is error-level and even constrained device may take advantage of error logging when debugging things. :)

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 4 above)
Comment (Damian-Nordic) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:355]: Imho, if the list is empty, then this function should return false. We shouldn't be using some stale entry.

Comment (hrajendran-samsung) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp:355]: I see your point about the stale entry. Will remove.

File: src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp (hunk 3 above)
Comment (selissia) [reply src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp:79]: Yes, TriggerImmediateQuery() should set mLastUsedProvider, it's a bug. It should be set every time SetCurrentProviderLocation() is called.

Comment (selissia) [reply src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp:79]: https://github.com/project-chip/connectedhomeip/issues/15691 will add another safeguard in this area

**Review by Damian-Nordic — APPROVED** [review]

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp (hunk 4 above)
Comment (selissia) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: Last used provider could be the one we received in the AnnounceOTAProvider command, it might not be present in the Default OTA Providers List and we don't add it to that list -- so this is a valid situation. 

Whether we want to keep retrying the provider from AnnounceOTAProvider is a separate question -- in my opinion either choice is fine, may as well retry.

Comment (Damian-Nordic) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp]: If we want to use the provider announced with `AnnounceOTAProvider` then I think we need to additionally mark the last used provider source, so that we don't use a removed default provider.

**Review by selissia — APPROVED** [review]

File: src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp [hunk 5]
```diff
@@ -68,9 +68,17 @@ void ExtendedOTARequestorDriver::PollUserConsentState()
...
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/ExtendedOTARequestorDriver.cpp:79]: Just noticed the function name being logged here as well

**Review by carol-apple — APPROVED** [review]

**Comment by hrajendran-samsung** [comment]
Squashed commits and rebased.

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.h [hunk 6]
```diff
@@ -87,7 +87,7 @@ class GenericOTARequestorDriver : public OTARequestorDriver
     uint32_t mPeriodicQueryTimeInterval = (24 * 60 * 60); // Timeout for querying providers on the default OTA provider list
 
     using ProviderLocationType = app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type;
-    ProviderLocationType mLastUsedProvider; // Provider location used for the last query or update
+    Optional<ProviderLocationType> mLastUsedProvider; // Provider location used for the last query or update
```
Comment (tcarmelveilleux) [inline src/app/clusters/ota-requestor/GenericOTARequestorDriver.h:90]: Since this is not stored, it means the first item in the list is always preferred on a cold boot, rather than last successful. Perhaps storing last successful entry to see mLastUsedProvider at init would be worthwhile. This can be done as a follow-up, since it's an optimization

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/clusters/ota-requestor/GenericOTARequestorDriver.h (hunk 6 above)
Comment (hrajendran-samsung) [reply src/app/clusters/ota-requestor/GenericOTARequestorDriver.h:90]: I think the idea would be to persist the lastUsedProvider in the flash, since the requestor will have to send the NotifyUpdateApplied message to this provider, post-installation of the new firmware. But yes, that would be a follow-up.
